onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/bias_softmax_fusion.h"
+
+#include <deque>
+
+#include "core/common/logging/logging.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+#include "core/providers/common.h"
+#include "core/graph/graph_utils.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+// helper to check dimensions match on concrete or symbolic value
+bool operator==(const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs, int value) {
+  return utils::HasDimValue(lhs) && lhs.dim_value() == value;
+}
+
+bool operator!=(const ONNX_NAMESPACE::TensorShapeProto_Dimension& lhs, int value) {
+  return !(lhs == value);
+}
+
+void select_input_on_lhs_condition(bool lhs_condition, Node& add_node, NodeArg*& input, NodeArg*& mask) {
+  if (lhs_condition) {
+    input = add_node.MutableInputDefs()[0];
+    mask = add_node.MutableInputDefs()[1];
+  }
+  else {
+    input = add_node.MutableInputDefs()[1];
+    mask = add_node.MutableInputDefs()[0];
+  }
+}
+
+// pattern match on dropout(softmax(input + mask)) subgraph
+bool TryBiasSoftmaxSubgraphMatch(Graph& graph, Node& start, Node*& add, Node*& softmax) {
+
+    Node& node = start; 
+    add = softmax = nullptr;
+
+    // check node is add and has single output
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Add", {7}) ||
+        !graph_utils::IsSupportedProvider(node, {kCudaExecutionProvider}) ||
+        !optimizer_utils::CheckOutputEdges(graph, node, 1)) {
+      return false;
+    }
+
+    // check shape information is not available for both add inputs
+    Node& add_node = node;
+    NodeArg* input1 = add_node.MutableInputDefs()[0];
+    NodeArg* input2 = add_node.MutableInputDefs()[1];
+    const TensorShapeProto* S1 = input1->Shape();
+    const TensorShapeProto* S2 = input2->Shape();
+    if (S1 == nullptr || S2 == nullptr || S1->dim_size() < 1 || S2->dim_size() < 1) {
+      return false;
+    }
+
+    // check add is only consumed by softmax with matching exec provider
+    auto p = add_node.OutputNodesBegin();
+    if (p == add_node.OutputNodesEnd()) {
+      return false;
+    }
+    Node& softmax_node = *graph.GetNode(p->Index());
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(softmax_node, "Softmax", {1, 11}) ||
+        softmax_node.GetExecutionProviderType() != add_node.GetExecutionProviderType()) {
+      return false;
+    }
+
+    // can't perform conversion if output is graph output
+    if (!graph.GetNodeOutputsInGraphOutputs(add_node).empty()) {
```

Comment:
this check and `if (p == add_node.OutputNodesEnd())` above seem redundant given the earlier call to `!optimizer_utils::CheckOutputEdges(graph, node, 1)`